src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,400 @@
+    variance = arma::mean(arma::mean(arma::pow(
+        inputTemp.each_slice() - arma::repmat(mean,
+        inputSize, 1), 2), 2), 0);
+
+    outputTemp.each_slice() -= arma::repmat(mean, inputSize, 1);
+
+    // Used in backward propagation.
+    inputMean.set_size(arma::size(inputTemp));
+    inputMean = outputTemp;
+
+    // Normalize output.
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(variance,
+        inputSize, 1) + eps);
+
+    // Re-used in backward propagation.
+    normalized.set_size(arma::size(inputTemp));
+    normalized = outputTemp;
+
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
+
+    count += 1;
+    // Value for average factor which used to update running parameters.
+    double averageFactor = average ? 1.0 / count : momentum;
+
+    double nElements = 0.0;
+    if (input.n_elem - size != 0)
+      nElements = 1.0 / (input.n_elem - size + eps);
+
+    // Update running mean and running variance.
+    runningMean = (1 - averageFactor) * runningMean + averageFactor *
+        mean.t();
+    runningVariance = (1 - averageFactor) * runningVariance +
+       input.n_elem * nElements *
+       averageFactor * variance.t();
+  }
+  else
+  {
+    // Normalize the input and scale and shift the output.
+    output = input;
+    arma::Cube<typename MatType::elem_type> outputTemp(
+        const_cast<MatType&>(output).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+
+    outputTemp.each_slice() -= arma::repmat(runningMean.t(),
+        inputSize, 1);
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(runningVariance.t(),
+        inputSize, 1) + eps);
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
+  }
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Backward(
+    const MatType& input,
+    const MatType& gy,
+    MatType& g)
+{
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = inputDimension;
+
+  arma::Cube<typename MatType::elem_type> gyTemp(
+      const_cast<MatType&>(gy).memptr(), inputSize, size,
+      batchSize * higherDimension, false, false);
+  arma::Cube<typename MatType::elem_type> gTemp(
+      const_cast<MatType&>(g).memptr(), inputSize, size,
+      batchSize * higherDimension, false, false);
+
+  // Step 1: dl / dxhat.
+  arma::Cube<typename MatType::elem_type> norm =
+      gyTemp.each_slice() % arma::repmat(gamma.t(), inputSize, 1);
+
+  // Step 2: sum dl / dxhat * (x - mu) * -0.5 * stdInv^3.
+  MatType temp = arma::sum(norm % inputMean, 2);
+  MatType vars = temp % arma::repmat(arma::pow(stdInv, 3),
+      inputSize, 1) * -0.5;
+
+  // Step 3: dl / dxhat * 1 / stdInv + variance * 2 * (x - mu) / m +
```

Comment:
Did a negative sign get dropped here? Shouldn't it be `-2` not `2`? (according to the paper)

Reply:
Yeah, I am confused by steps 3 and 4 here. I suspect this may be where the bug is. I would have expected us to compute `dl/dmu` in step 3, then `dl/dx` in step 4, just like in the paper.  But instead step 3 kind of looks like the equation for `dl/dx`, and step 4 looks like half the equation for `dl/dmu`.  It seems like this line uses `(inputMean.each_slice() % vars * 2)) / (batchSize * higherDimension)` for `dl/dmu`, but as far as I can tell that is not correct.  Maybe it makes sense to just try and rewrite these two steps according directly to the paper?